### PR TITLE
fluxv2 plugin: incremental (cache refactoring, misc bug fixes and TODOs)

### PR DESCRIPTION
**Pull request by gfichtenholt**
No new features in this PR. Mainly:
- fixing bugs
- miscellaneous TODOs left over from previous PRs 
- slightly different handling of errors, using runtime.HandleError instead of log.Error where applicable
- a pretty big refactoring of the current caching code to introduce a go ratelimiting queue to process events in order avoid race conditions. Also added an integration test for this scenario

before I embark on #3636

Update 11/19: This is ready for review now. This has been the most challenging PR I have done yet. The main problem I was solving here were concurrency-related issues in the cache. Let me give a quick summary at how I arrived where I am:
1) the cache mainly works by registering a k8s watcher and watching for events. Lets call it a "sink". The processing of the events must be somewhat quick, or else the sink might get flooded. To solve this, I used to  have code like `go c.onModify() `, in other words, invoking the appropriate handler in a goroutine. As I discovered over a week ago, this was problematic and could lead to race conditions because you may have concurrent writers updating the same cache entry and leave the cache in an inconsistent state
2) To solve this, I decided to follow the same pattern as our apprepository-controller (in fact, the pattern is common to [many controllers](https://engineering.bitnami.com/articles/a-deep-dive-into-kubernetes-controllers.html)) of using workqueue.RateLimitingInterface. Upon receiving the event, I place a key on the queue, and "worker" running in the background will pull it off the queue and sync the corresponding k8s object with the redis cache. So far so good.
3) It turned out that in addition to the pattern mentioned in (2), there was another completely pattern of accessing the cache. For example, a call go GetAvailablePackageVersions() needs to get the contents of the cache entry (parsed index model). The entry may not necessarily be there. For example, the entry may not be there if it has been evicted previously by redis in order to make room for some other entry. So in this case, I need to go back and fetch the corresponding k8s object, process it and add it back to the cache, possibly evicting some other entry in the process. So this is another kind of cache writer, different from that of the sink. Now imagine 2 or more of these kind of writers concurrently executing. If I just let them write to the cache directly, I again will run into race conditions. So I need to "funnel" these kinds of writers and at the same time make sure it doesn't interfere with the sink writer. This leads me to where I am today - instead of writing to the cache entry directly, the code will place an object in the same queue as (2) and wait until the worker process is finished processing this entry. It turned out that the implementation of workqueue.RateLimitingInterface in k8s in was lacking one feature (the 'WaitUntilDoneProcessing' func) so I rolled my own, based on their code.

I do realize this is a difficult PR to review, as there are many changes, so take your time. I am working on things in parallel. I am also available for a zoom session to discuss if you would prefer

### CODE REVIEW

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/integration_utils_test.go
Comment (absoludity) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/integration_utils_test.go:201]: (that is, in my own branch)

**Review by absoludity — APPROVED** [review]
> No new features in this PR. Mainly:
> 
>     * fixing bugs
> 
>     * miscellaneous TODOs left over from previous PRs
> 
>     * slightly different handling of errors, using runtime.HandleError instead of log.Error where applicable
> 
>     * a pretty big refactoring of the current caching code to introduce a go ratelimiting queue to process events in order avoid race conditions. Also added an integration test for this scenario
> 
> 
> before I embark on #3636

[snip]

> 
> I do realize this is a difficult PR to review, as there are many changes, so take your time. I am working on things in parallel. I am also available for a zoom session to discuss if you would prefer

Thanks Greg. I appreciate all the comments and explanations that you've included. Yes, a diff at  +1,669 −911 of concurrency code is pretty difficult for me to review it properly. I've tried but my concentration waned after a while. For the future, what I'd personally prefer is PRs that have a tightly focused change around a single concept. It's hard sometimes I know, but in this case, the four topics you mention above could have been a good split to enable a reviewer to focus on each (possibly with the last one being split up into your cache implementation, then the update to use it).

Anyway, +1, given that this is a plugin. The one thing that does worry me a little (mentioned inline) is that the caching code is getting quite complex. This may be unavoidable given the problems that you've found, I just worry about having code that depends on any one person's knowledge and understanding. Hopefully when we eventually generalize the caching code for other plugins we may have a chance for different members to familiarize and work on the code, and the great comments you've included throughout will be very helpful!

**Comment by gfichtenholt** [comment]
> > No new features in this PR. Mainly:
> > ```
> > * fixing bugs
> > 
> > * miscellaneous TODOs left over from previous PRs
> > 
> > * slightly different handling of errors, using runtime.HandleError instead of log.Error where applicable
> > 
> > * a pretty big refactoring of the current caching code to introduce a go ratelimiting queue to process events in order avoid race conditions. Also added an integration test for this scenario
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > before I embark on #3636
> 
> [snip]
> 
> > I do realize this is a difficult PR to review, as there are many changes, so take your time. I am working on things in parallel. I am also available for a zoom session to discuss if you would prefer
> 
> Thanks Greg. I appreciate all the comments and explanations that you've included. Yes, a diff at +1,669 −911 of concurrency code is pretty difficult for me to review it properly. I've tried but my concentration waned after a while.

yes, I understand. I tried my best explaining what it is I was solving, but even so, the end result is complex for a person unfamiliar with the details to keep track of. I get that.

> For the future, what I'd personally prefer is PRs that have a tightly focused change around a single concept. It's hard sometimes I know, but in this case, the four topics you mention above could have been a good split to enable a reviewer to focus on each (possibly with the last one being split up into your cache implementation, then the update to use it).

Understood. Will do my best in the future. For what its worth, even with this PR, it didn't start out that way. Most of the changes had to do with solving a single race condition in the cache, which I originally thought would be not be complex to solve. But it turned out like peeling an onion: in order to have X, you have to have done Y, and in order to have Y, you have to have done X, etc. Over the course of a 2 weeks it just mushroomed into what you see today. I didn't want to push a half-assed solution
> 
> Anyway, +1, given that this is a plugin. 
Thanks

> The one thing that does worry me a little (mentioned inline) is that the caching code is getting quite complex. This may be unavoidable given the problems that you've found, I just worry about having code that depends on any one person's knowledge and understanding.

Yes, I share your concern. I am happy to my part and go over the plot  in general any part specifically over zoom, emails, slack or any other communication method,

> Hopefully when we eventually generalize the caching code for other plugins we may have a chance for different members to familiarize and work on the code, and the great comments you've included throughout will be very helpful!
Yes, that was my hope too. Part of this PR was moving the cache code, which is generic, not specific to flux out to a separate package. That ended up pretty nicely, as it really forced my to think about what functions and state to expose to caller an which can be kept internally. Not to toot my pwn horn, but I'm quite proud of where I ended up. Seems pretyy clean to me and provides a sl=o

lid foundation

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/chart_cache.go
Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/chart_cache.go]: Yes, it's run as part of ```
func RegisterWithGRPCServer(s grpc.ServiceRegistrar, configGetter core.KubernetesConfigGetter,
	clustersConfig kube.ClustersConfig, pluginConfigPath string) (interface{}, error) {
```
So you're saying we don't want to return grpc statuses in this context? OK with me. I will fix.

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/integration_utils_test.go
Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/integration_utils_test.go:190]: yikes, that's like a 100 places. Thanks :-)
Will do

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/integration_utils_test.go
Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/integration_utils_test.go:207]: yep, seems to be the case. Sorry, I forget where the original place where I got the idea. Possibly official k8s example(s) for watches?

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/release_integration_test.go
Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/release_integration_test.go:649]: well, not quite. Unless you meant "end as soon as it was NOT zero". Here I am testing that I *DON'T* get any events. In other words, I don't want any events to come in. Ever. So if I could wait for an hour, I would :-). The expected behavior is NOT to get any events, checking every 5s isn't going to improve much when things go as expected. So bleh. I dislike using time.Sleep() as much as the next guy. If I think of something better I will change it

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/release_integration_test.go
Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/release_integration_test.go:603]: yep, took me two days to make it work :-)

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/watcher_cache.go
Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/watcher_cache.go]: my pleasure. I always try with the idea someone else might have to jump in and maintain the code.

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/watcher_cache.go
Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/watcher_cache.go]: yep, 2nd time you bring this up :-). Don't worry about this. I will remove it in time, I promise.

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache/rate_limiting_queue.go
Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache/rate_limiting_queue.go:47]: yes, this was quite a challenge that took several days to debug and get right. The code was based on https://github.com/kubernetes/client-go/blob/v0.22.4/util/workqueue/queue.go and https://github.com/kubernetes/client-go/blob/v0.22.4/util/workqueue/rate_limiting_queue.go  but tailored a bit to my needs. I needed a couple extra methods that accessed queue internal state. The use of rateLimitingType struct is exactly what the original source was doing https://github.com/kubernetes/client-go/blob/v0.22.4/util/workqueue/rate_limiting_queue.go#L51. The rateLimiter field is there just so I don't have to write code like 
`workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "AppRepositories"),`
(save some typing). If you follow through the code, you should be able to see the same things are happening: there is a  rateLimiterType object with a queue and a rateLimiter. Mine just happens to hardcode what rateLimiter is (for now)

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache/rate_limiting_queue.go
Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache/rate_limiting_queue.go:81]: I didn't choose the name :-)
https://github.com/kubernetes/client-go/blob/29f81448d77a50b041e21e44cbf0269450e8ae03/util/workqueue/queue.go#L71
I chose to copy and paste :-)

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache/rate_limiting_queue.go
Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache/rate_limiting_queue.go:204]: yes, indeed. I was on quite a learning curve over the last 2 weeks while working on this PR. Did I mention it was challenging? :-)

File: cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache/watcher_cache.go
Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache/watcher_cache.go:873]: Geez I feel like going through an interview process where the interviewer asks "what do you do when a team member disagrees with you or questions your choices". Yes, you mentioned this already Michael. I don't know whether this is a criticism or what. I don't quite know how to respond to make you happy. I too, wish things were simple. Having said that, this particular for loop follows exactly the pattern the documentation tells you to do:
https://pkg.go.dev/sync#Cond.Wait
In general, I don't set out to write complicated code. If in the end it happens to be complicated, it's just how it evolved in order to solve some problem.  When have a problem in front of me, I try to solve it the best way I can. Any constructive ideas, such as how to solve a problem in a simpler way, I am generally open to.  Thanks

Comment (absoludity) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache/watcher_cache.go:895]: Sorry Greg. I don't mean to leave you feeling that way, nor did I mean to mention it more than once. Sometimes I write comments inline then summarize when submitting the review. The comment here is the inline one when I found at the end of the file that I was really struggling to stay focused any more, but I kept going to the end, and added the review summary. I'll update to stop adding a summary so it doesn't get repeated.

Nor am I trying to disagree with your choices here... as above, it's excellent that you test the code so well as that makes it much safer for anyone to work on, which is great. I am *very* happy that we've got someone who dives in to an issue like this with the goal of finding the best way to solve it, it's excellent. Sometimes a problem is complex and the solution will reflect that, of course.

In terms of reviewing large complex diffs, yep, it's sometimes inevitable, and when they are, it's hard to do a good job as a reviewer (I feel it's hard for me at least). One constructive idea that I find helpful for myself is that of "prequel" (rather than sequel) branches: if I'm working on a PR to do X, but while working on X I find I also need to do Y to be able to do X, where possible (and it's not always), I save my work, create a new branch from main, do Y, push that for review, rebase my X branch on that one and continue while I wait for Y to be reviewed and landed. Once Y is landed I can rebase on main and continue. Not saying that's possible here, nor that it's something you don't do or have not done in the past, just something I find helpful. (Here's an [example prequel PR](https://github.com/kubeapps/kubeapps/pull/3781)).

The day was also a long day of reviews (and only reviews) for me, which was pretty hard to get through. I'll try to make sure next time that I don't just review all day, but instead just set part of the day aside and do them over a couple of days when there are so many.

Comment (gfichtenholt) [reply cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache/watcher_cache.go:895]: I understand. Thanks for all your comments. Cheers

**Pull request merged**
